src/sqlancer/datafusion/DataFusionProvider.java

```diff
@@ -0,0 +1,135 @@
+package sqlancer.datafusion;
+
+import static java.lang.System.exit;
+import static sqlancer.datafusion.DataFusionUtil.DataFusionLogger.DataFusionLogType.DML;
+import static sqlancer.datafusion.DataFusionUtil.displayTables;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import com.google.auto.service.AutoService;
+
+import sqlancer.DatabaseProvider;
+import sqlancer.IgnoreMeException;
+import sqlancer.Randomly;
+import sqlancer.SQLConnection;
+import sqlancer.SQLGlobalState;
+import sqlancer.SQLProviderAdapter;
+import sqlancer.common.query.SQLQueryAdapter;
+import sqlancer.datafusion.DataFusionProvider.DataFusionGlobalState;
+import sqlancer.datafusion.DataFusionSchema.DataFusionTable;
+import sqlancer.datafusion.DataFusionUtil.DataFusionInstanceID;
+import sqlancer.datafusion.DataFusionUtil.DataFusionLogger;
+import sqlancer.datafusion.gen.DataFusionInsertGenerator;
+import sqlancer.datafusion.gen.DataFusionTableGenerator;
+
+@AutoService(DatabaseProvider.class)
+public class DataFusionProvider extends SQLProviderAdapter<DataFusionGlobalState, DataFusionOptions> {
+
+    public DataFusionProvider() {
+        super(DataFusionGlobalState.class, DataFusionOptions.class);
+    }
+
+    @Override
+    public void generateDatabase(DataFusionGlobalState globalState) throws Exception {
+        int tableCount = Randomly.fromOptions(1, 2, 3, 4, 5, 6, 7);
+        for (int i = 0; i < tableCount; i++) {
+            SQLQueryAdapter queryCreateRandomTable = new DataFusionTableGenerator().getQuery(globalState);
+            queryCreateRandomTable.execute(globalState);
+            globalState.updateSchema();
+            globalState.dfLogger.appendToLog(DML, queryCreateRandomTable.toString() + "\n");
+        }
+
+        // Now only `INSERT` DML is supported
+        // If more DMLs are added later, should use`StatementExecutor` instead
+        // (see DuckDB's implementation for reference)
+
+        globalState.updateSchema();
+        List<DataFusionTable> allTables = globalState.getSchema().getDatabaseTables();
+        List<String> allTablesName = allTables.stream().map(t -> t.getName()).collect(Collectors.toList());
+        if (allTablesName.isEmpty()) {
+            System.out.println("Generate database failed");
+            exit(1);
+        }
+
+        // Randomly insert some data into existing tables
+        for (DataFusionTable table : allTables) {
+            int nInsertQuery = globalState.getRandomly().getInteger(0, 8); // [0, 10)
```

Comment:
Here, you could consider using the parameter that also other SQLancer implementations use (see https://github.com/sqlancer/sqlancer/blob/852c1e4fecc853dc6246b45a3cc3045da9d8e920/src/sqlancer/MainOptions.java#L30). It's fine to do this as a follow-up PR if you want, perhaps also while adopting the `StatementExecutor` class used by other implementations).